framework/src/main/java/org/checkerframework/framework/ajava/ExpectedTreesVisitor.java

```diff
@@ -0,0 +1,421 @@
+                MethodTree methodTree = (MethodTree) member;
+                if (methodTree.getName().contentEquals("<init>")) {
+                    continue;
+                }
+            }
+
+            member.accept(this, p);
+        }
+
+        return null;
+    }
+
+    @Override
+    public Void visitLambdaExpression(LambdaExpressionTree tree, Void p) {
+        for (VariableTree parameter : tree.getParameters()) {
+            // Parameter types might not be specified for lambdas. When not specified, JavaParser
+            // uses UnknownType. Conservatively, don't add parameter types for lambda expressions.
+            visit(parameter.getModifiers(), p);
+            visit(parameter.getNameExpression(), p);
+            assert parameter.getInitializer() == null;
+        }
+
+        visit(tree.getBody(), p);
+        return null;
+    }
+
+    @Override
+    public Void visitWhileLoop(WhileLoopTree tree, Void p) {
+        super.visitWhileLoop(tree, p);
+        // javac surrounds while loop conditions in a ParenthesizedTree but JavaParser does not, so
+        // don't check the condition tree.
+        trees.remove(tree.getCondition());
+        return null;
+    }
+
+    @Override
+    public Void visitDoWhileLoop(DoWhileLoopTree tree, Void p) {
+        super.visitDoWhileLoop(tree, p);
+        // javac surrounds while loop conditions in a ParenthesizedTree but JavaParser does not, so
+        // don't check the condition tree.
+        trees.remove(tree.getCondition());
+        return null;
+    }
+
+    /**
+     * Calls the correct visit method for {@code tree} if {@code tree} is non-null.
+     *
+     * @param tree the tree to visit
+     * @param p secondary parameter to the visitor
+     */
+    private void visit(@Nullable Tree tree, Void p) {
+        if (tree != null) {
+            tree.accept(this, p);
+        }
+    }
+
+    /**
+     * If {@code trees} is non-null, visits each non-null tree in {@code trees} in order.
+     *
+     * @param trees the list of trees to visit
+     * @param p secondary parameter to the visitor
+     */
+    private void visit(@Nullable List<? extends @Nullable Tree> trees, Void p) {
+        if (trees == null) {
+            return;
+        }
+
+        for (Tree tree : trees) {
+            visit(tree, p);
+        }
+    }
+
+    /** Visitor that records whether it has visited a "var" keyword. */
+    private static class HasVarTypeVisitor extends VoidVisitorAdapter<Void> {
+        /** Whether a "var" keyword has been visited. */
+        public boolean hasVarType;
+
+        /** Constructs a visitor that hasn't yet recorded visiting a "var" keyword. */
+        public HasVarTypeVisitor() {
+            hasVarType = false;
+        }
+
+        @Override
+        public void visit(ClassOrInterfaceType node, Void p) {
+            if (node.getName().asString().equals("var")) {
+                hasVarType = true;
+            }
+
+            super.visit(node, p);
```

Comment:
Once `hasVarType` is true, there is no need to do any more visiting.

(For efficiency, it would even be possible to shortcut, such as by throwing an exception that could be caught by the static `hasVarType` method that I added, and then you also wouldn't need the variable.  You could skip this optimization if you find it gross.)

Reply:
This class is only used by the testing framework, and this would only be an optimization if the file actually uses `var`, which doesn't occur often in the tests as far as I know.